File: scripts/check_static_rules.mjs
import assert from 'node:assert/strict';
import {readFileSync} from 'node:fs';
import * as rules from '../web/games/paths-of-glory.js';
crypto.getRandomValues=array=>{array.fill(0);return array;};
const cases=JSON.parse(readFileSync(process.argv[2],'utf8'));
for(const [i,c] of cases.entries()){
 const original=structuredClone(c.before);
 try {
 if(c.error)assert.throws(()=>rules.apply(c.before,c.action));
 else {const actual=rules.apply(c.before,c.action);assert.deepEqual(actual.state,c.after);assert.equal(actual.message,c.message);assert.deepEqual(rules.publicView(actual.state),c.public);}
 assert.deepEqual(c.before,original,'Input state must remain unchanged');
 }catch(error){throw new Error(`Parity case ${i}: ${JSON.stringify(c.action)}\n${error.stack}`);}
}
for(const max_hand of [7,8])for(const guns of [true,false]){const s=rules.create({max_hand,guns});rules.validate(s);assert.equal(s.sides.cp.deck.length,10);assert.equal(s.sides.cp.remaining,max_hand);assert.equal(s.pending?.forced,guns?'guns':undefined);}
console.log(`PASS: ${cases.length} Python/browser transitions, messages, candidates and public projections; 4 setup variants.`);

File: scripts/check_static_rules.py
"""Replay the Python rules suite's transitions through the browser adapter using Node.
Randomness is fixed for both engines; state, messages, legal actions and privacy agree.
"""
import copy
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile
import unittest
from unittest.mock import patch

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))
sys.path.insert(0, str(ROOT / 'tests'))
from cdg import engine
from cdg.games import paths_of_glory as g
from test_rules import RulesTests

original = engine.apply
cases = []

def capture(state, action):
    case = {'before': copy.deepcopy(state), 'action': copy.deepcopy(action)}
    try:
        updated, message = original(state, action)
    except ValueError:
        case['error'] = True
        cases.append(case)
        raise
    case.update(after=copy.deepcopy(updated), message=message, public=engine.view(updated))
    # Browser asset URLs are relative so /repo/ works on GitHub Pages.
    for side in case['public']['sides'].values():
        for slot in side['display'].values():
            if slot and slot['card']:
                slot['card']['image'] = slot['card']['image'].lstrip('/')
    cases.append(case)
    return updated, message

with patch.object(engine, 'apply', capture), patch.object(g, 'shuffled', lambda ids: list(ids)[1:] + list(ids)[:1]), patch.object(g.secrets, 'randbelow', lambda n: 0):
    result = unittest.TextTestRunner(verbosity=0).run(unittest.defaultTestLoader.loadTestsFromTestCase(RulesTests))
if not result.wasSuccessful():
    raise SystemExit(1)

with tempfile.TemporaryDirectory() as temporary:
    fixtures = Path(temporary) / 'cases.json'
    fixtures.write_text(json.dumps(cases, ensure_ascii=False))
    subprocess.run([os.environ.get('NODE', 'node'), str(ROOT / 'scripts/check_static_rules.mjs'), str(fixtures)], check=True)
